dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ExternalSourceRestRepository.java

```diff
@@ -0,0 +1,109 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.app.rest.repository;
+
+import java.util.List;
+import java.util.Optional;
+
+import org.dspace.app.rest.converter.ExternalSourceEntryRestConverter;
+import org.dspace.app.rest.converter.ExternalSourceRestConverter;
+import org.dspace.app.rest.model.ExternalSourceEntryRest;
+import org.dspace.app.rest.model.ExternalSourceRest;
+import org.dspace.external.model.ExternalDataObject;
+import org.dspace.external.provider.ExternalDataProvider;
+import org.dspace.external.service.ExternalDataService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.rest.webmvc.ResourceNotFoundException;
+import org.springframework.stereotype.Component;
+
+/**
+ * This is the Repository that is responsible for the functionality and implementations coming from
+ * {@link org.dspace.app.rest.ExternalSourcesRestController}
+ */
+@Component(ExternalSourceRest.CATEGORY + "." + ExternalSourceRest.NAME)
+public class ExternalSourceRestRepository extends AbstractDSpaceRestRepository {
```

Comment:
I see a lot of (conceptual) similarities between the `/api/integrations/externalsources` endpoint (represented here) and the existing `/api/integrations/authorities` endpoint.  So, I'm slightly surprised that this `ExternalSourceRestRepository` (and the corresponding Controller) seems to take a different implementation approach to that of `AuthorityRestRepository`, `AuthorityEntryLinkRepository` and `AuthorityEntryValueLinkRepository` (the three of which work together so there is no `AuthorityRestController`).   Could you review how these Authority classes implement something similar and see if it can be applied here?   I worry we are building a lot of reliance on custom Controllers to override default behaviors, when we could be using RestRepository (and LinkRestRepository) classes to do the same thing.

I'd like us to follow existing practices where possible. Since these endpoints behave similarly, I wonder if we can use the same approach to both?

Reply:
@tdonohue I agree with the premise that we should attempt to limit the number of controllers here. But the way it is currently implemented in the authority rest side is not the best approach for this functionality. It's hard to read and has a high risk of breaking the functionality with changes. It will be hard for future developers to find out how it works in the first place.

**How it works now**

The AuthorityEntryLinkRepository has a single “query” method while the AuthorityEntryValueLinkRepository has a single “getResource” method. When using intellij to find the usages of these methods, none are found. So it isn’t immediately clear how these are called in the first place. After some searching we found that the following controller method is responsible for the call: https://github.com/DSpace/DSpace/blob/master/dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java#L777
Reflection is used here to find the correct repository, the actual method call depends on the attributes found in the rest class: https://github.com/DSpace/DSpace/blob/master/dspace-server-webapp/src/main/java/org/dspace/app/rest/model/AuthorityRest.java#L17
Should you rename one of the methods in these repositories, it would still compile your code, but you would encounter exceptions at runtime. This is something we should avoid.

**Moving forward**

Further decreasing the controllers is perhaps something to verify next since the first iteration of projections is in place, and the next step there can be a detailed analysis on reflection and the RestResourceController to support permissions, .... Once the RestResourceController is refactored, it can be much easier to work with e.g. beans which will allow developers to more easily find out how repositories are contacted. But this is something that we would have to do in a separate PR dedicated to that single refactor.

Reply:
Thanks for the detailed response here, @KevinVdV .  The direction you've chosen sounds fine to me, and I'm good with this PR's approach as it has been recently refactored. Marking this conversation as "resolved"